presto/serverprivate.nim

```diff
@@ -107,230 +107,255 @@ proc processRestRequest*[T](
+      try:
+        await request.getContentBody()
+      except HttpTransportError as exc:
+        debug "Unable to receive request body", uri = $request.uri,
+              peer = $request.remoteAddress(), meth = $request.meth,
+              error_msg = $exc.msg
+
+        when defined(metrics):
+          processStatusMetrics(route, Http400)
+
+        return sresponse(request, Http400,
+                         RestRequestError.InvalidContentBody)
+      except HttpProtocolError as exc:
+        debug "Unable to obtain request body", uri = $request.uri,
+              peer = $request.remoteAddress(), meth = $request.meth,
+              error_msg = $exc.msg
+
+        when defined(metrics):
+          processStatusMetrics(route, Http400)
+
+        return sresponse(request, Http400,
+                         RestRequestError.InvalidContentBody)
+      except RestBadRequestError as exc:
+        debug "Request has incorrect content type", uri = $request.uri,
+               peer = $request.remoteAddress(), meth = $request.meth,
+               error_msg = $exc.msg
+
+        when defined(metrics):
+          processStatusMetrics(route, Http400)
+
+        return sresponse(request, Http400,
+                         RestRequestError.InvalidContentType)
     else:
-      debug "Received invalid request", peer = $request.remoteAddress(),
-            meth = $request.meth, uri = $request.uri
+      none[ContentBody]()
+
+  when T isnot RestServerMiddlewareRef:
+    debug "Serving API request", peer = $request.remoteAddress(),
+          meth = $request.meth, uri = $request.uri,
+          path_params = pathParams, query_params = queryParams,
+          content_body = optBody
+
+  when defined(metrics):
+    let responseStart = Moment.now()
+
+  let
+    restRes =
+      block:
+        let res = await route.callback(request, pathParams, queryParams,
+                                       optBody)
+        when defined(metrics):
+          processMetrics(route, Moment.now() - responseStart)
+
+        res
+
+  if request.responded():
+    return block:
+      when T isnot RestServerMiddlewareRef:
+        debug "Response was sent in request handler", meth = $request.meth,
+              peer = $request.remoteAddress(), uri = $request.uri,
+              path_params = pathParams, query_params = queryParams,
+              content_body = optBody
+      defaultResponse()
+
+  try:
+    case restRes.kind
+    of RestApiResponseKind.Empty:
+      when T isnot RestServerMiddlewareRef:
+        debug "Received empty response from handler",
+                meth = $request.meth, peer = $request.remoteAddress(),
+                uri = $request.uri
 
       when defined(metrics):
-        presto_server_invalid_requests_count.inc()
+        processStatusMetrics(route, Http410)
+
+      await request.respond(Http410)
+
+    of RestApiResponseKind.Status:
+      var headers = HttpTable.init()
+      if server.router.allowedOrigin.isSome:
```

Comment:
Forgive me to put my request here as its easier to point out.
This is the place which I would like to have it decoupled from the core processing of rest requests.
As it is ok that I can chain rest request processors as middlewares but having allow-origin handled here I will not be (or just with unnecessage response header hacking) to separately influence this logic.
It woulbe beneficial if I can use the rest middleware where I can put my CORS logic separately, ideally in a separate middleware and injected into the chain.
WDYT is it solveable?

Reply:
You can disable internal PRESTO CORS mechanism if you create `RestRouter` instance with supplying `allowedOrigin` argument.